scaper/core.py

```diff
@@ -1913,29 +1986,94 @@ def _generate_audio(self, audio_path, ann, reverb=None,
+
+                if peak_normalization or (clipping and fix_clipping):
+
+                    # normalize soundscape audio and scale event audio
+                    soundscape_audio, event_audio_list, scale_factor = \
+                        peak_normalize(soundscape_audio, event_audio_list)
+
+                    ref_db_change = 20 * np.log10(scale_factor)
+
+                    if clipping and fix_clipping:
+                        warnings.warn(
+                            'Peak normalization applied to fix clipping with '
+                            'scale factor = {}. The actual ref_db of the '
+                            'generated soundscape audio will change by '
+                            'approximately {:.2f}dB with respect to the target '
+                            'ref_db of {})'.format(
+                                scale_factor, ref_db_change, self.ref_db),
+                            ScaperWarning)
+
+                    if scale_factor < 0.05:
+                        warnings.warn(
+                            'Scale factor for peak normalization is extreme '
+                            '(<0.05), event SNR values in the generated soundscape '
+                            'audio may not perfectly match their specified values.',
+                            ScaperWarning
+                        )
+
+                # Apply effects and reshape
+                soundscape_audio = tfm.build_array(
+                    input_array=soundscape_audio,
+                    sample_rate_in=self.sr,
+                )
+                soundscape_audio = soundscape_audio.reshape(-1, self.n_channels)
+
+                # Optionally save soundscape audio to disk
+                if audio_path is not None:
+                    soundfile.write(audio_path, soundscape_audio, self.sr,
+                                    subtype='PCM_32')
+
+                # Optionally save isolated events to disk
                 if save_isolated_events:
                     base, ext = os.path.splitext(audio_path)
                     if isolated_events_path is None:
                         event_folder = '{:s}_events'.format(base)
                     else:
                         event_folder = isolated_events_path
 
-                    _role_count = role_counter[e.value['role']]
-                    event_audio_path = os.path.join(
-                        event_folder, 
-                        '{:s}{:d}_{:s}{:s}'.format(
-                            e.value['role'], _role_count, e.value['label'], ext))
-                    role_counter[e.value['role']] += 1
-                    
                     if not os.path.exists(event_folder):
-                        # In Python 3.2 and above we could do 
+                        # In Python 3.2 and above we could do
                         # os.makedirs(..., exist_ok=True) but we test back to
                         # Python 2.7.
                         os.makedirs(event_folder)
```

Comment:
The comment here suggests that we could get rid of the `if` block, as we now only support Python 3.5+.

Reply:
updated